runtime/lua/vim/net.lua

```diff
@@ -0,0 +1,533 @@
+    http_version = extra.http_version and tonumber(extra.http_version) or nil,
+    _raw_write_out = extra,
+  }
+end
+
+--- Asynchronously make network requests.
+---
+--- See man://curl for supported protocols. Not all protocols are fully tested.
+---
+--- Please carefully note the option differences with |vim.net.download()|, notably
+--- `redirect`.
+---
+---@see |vim.net.download()|
+---@see https://developer.mozilla.org/en-US/docs/Web/HTTP/Status
+---@see |job-control|
+---@see man://curl
+---
+---@param url string The request URL.
+---@param opts table|nil Optional keyword arguments:
+---             - method string|nil HTTP method to use. Defaults to GET.
+---             - headers HeaderTable | table<string, string | string[]> | nil Headers to set on the request
+---             - redirect string|nil Redirect mode. Defaults to "follow". Possible values are:
+---                 - "follow": Follow all redirects incurred when fetching a resource.
+---                 - "error": Throw an error using on_err or vim.notify when status is 3XX.
+---             - upload_file string|nil Path to an upload_file. Can be relative.
+---             - user string|nil String in "username:password" format. Prefer over passing in url.
+---             - data string|table|nil Data to send with the request. If a table, it will be
+---             JSON-encoded. vim.net does not currently support form encoding.
+---             - on_complete fun(response: table)|nil Callback function when request is
+---             completed successfully. The response has the following keys:
+---                 - ok boolean Whether the request was successful (status within 2XX range).
+---                 - headers fun(): HeaderTable Function returning a HeaderTable of response headers.
+---                 - text fun(): string|nil Function returning response body. If method was HEAD,
+---                 this is nil.
+---                 - json fun(opts: table|nil): table|nil Read the body as JSON. Optionally accepts
+---                 opts from |vim.json.decode|. Will throw errors if body is not JSON-decodable.
+---                 Nil if method is HEAD.
+---                 - method string The http method used in the most recent HTTP request.
+---                 - status number The numerical response code.
+---                 - size number The total amount of bytes that were downloaded. This
+---                 is the size of the body/data that was transferred, excluding headers.
+---                 - http_version number HTTP version used in the request.
+---             - on_err fun(err: string[], exit_code: number|nil)|nil Function recieving a `stderr_buffered` string[] of error.
+---             err is either curl stderr or internal fetch() error. Without providing this
+---             function, |vim.net.fetch()| will automatically raise the error to the user.
+---             See |on_stderr| and `stderr_buffered`.
+---@return number jobid A job id.
```

Comment:
I'm not sure if it's a good idea to expose the job id and refer to curl. It will make it more difficult to change the implementation later. E.g if we wanted to avoid spawning a process